flare/framework/manager.py

```diff
@@ -96,19 +95,8 @@ def start(self):
             if all([a.running.value for a in self.agents]):
                 break
 
-        while True:
-            # keep querying agents' progress
-            # whichever agent ends first, we end all
-            end_flag = False
-            for a in self.agents:
-                if not a.running.value:
-                    end_flag = True
-            if end_flag:
-                for a in self.agents:
-                    a.terminate()
-                break
-            else:
-                time.sleep(1.0)
+        for cdp in self.CDPs.values():
+            cdp.run([a.running for a in self.agents])
```

Comment:
`cdp` has to run before `agent`, otherwise it may miss the first few data that `agent` sends